net/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+//! holochain_net is a library that defines an abstract networking layer for
+//! different network transports and implements a PeerStore for mapping and
+//! managing the topology of transport layers with regard to relay's e.g. for NAT
+
+#[macro_use]
+extern crate failure;
+
+use failure::Error;
+
+pub mod error;
+
+pub type SerializedAddress = Vec<u8>;
+pub type TransportAddress = Vec<u8>;
+pub type SerializedMessage = Vec<u8>;
+
+/// this closure type will get called when the send completes and the parameter will be the response message (or error)
+type SendResponseClosure = Box<FnMut(Result<SerializedMessage, Error>) -> Option<Error> + Send>;
+
+/// this closure type gets called when a new message arrives, you can respond with a Message or an error.
+type ReceiveClosure =
+    Box<FnMut(&SerializedAddress, &SerializedMessage) -> Result<SerializedMessage, Error> + Send>;
+
+/// trait to encapsulate a node that represents a given Holochain app from the Transport's perspective
+pub trait Node {
+    fn get_address(&self) -> &SerializedAddress;
+    fn get_transport_address(&self) -> TransportAddress;
+}
+
+pub trait Transport {
+    /// initialize the transport
+    /// this might be used for example in a TCP based transport to specify a listening port for
+    /// for the transport.
+    fn initialize(&mut self, config: Option<String>) -> Result<(), Error>;
+
+    /// return a default config with best configuration guesses for the transport
+    fn get_default_config(&self) -> String;
+
+    /// register a peer as a node in the transport
+    fn new_node(
+        &mut self,
+        addr: SerializedAddress,
+        handler: Option<ReceiveClosure>,
+    ) -> Result<(), Error>;
+
+    /// send a message to a node over the transport
+    /// assumes that the sending address was registered locally with new_node
+    fn send(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        msg: SerializedMessage,
+        callback: SendResponseClosure,
+    ) -> Result<(), Error>;
+
+    /// deliver a message to a given recipient
+    /// assumes that recipient address was registered locally with new_node
+    /// this function may be called by the routing layer to bridge across transports
+    fn deliver(
+        &mut self,
+        from: &SerializedAddress,
+        to: &SerializedAddress,
+        message: SerializedMessage,
+    ) -> Result<SerializedMessage, Error>;
```

Comment:
Semantic, but in the ipc code, I had ended up calling the fire-and-forget message type `send`, and the procedure call style message type `call`. Anyone have quick vote, or should we name department this?

(p.s. I liked `send` and `call` because they are both four letters)